hal/src/dma.rs

```diff
@@ -0,0 +1,1075 @@
+}
+
+pub struct InterruptState {
+    /// Bitmask of allocated channels
+    channel_mask: VolatileCell<u32>,
+
+    /// IRQ critical section counter
+    cpu_irq_critical_section_counter: VolatileCell<u32>,
+
+    /// IRQ previous interrupt state
+    cpu_irq_prev_interrupt_state: VolatileCell<u8>,
+}
+
+impl Default for InterruptState {
+    fn default() -> Self {
+        Self {
+            channel_mask: VolatileCell::new(0),
+            cpu_irq_critical_section_counter: VolatileCell::new(0),
+            cpu_irq_prev_interrupt_state: VolatileCell::new(0),
+        }
+    }
+}
+
+/// Enter IRQ critical section
+fn cpu_irq_enter_critical(int_state: &InterruptState) {
+    atomic::compiler_fence(Ordering::Acquire);
+
+    if int_state.cpu_irq_critical_section_counter.get() == 0 {
+        if get_primask() == 0 {
+            // IRQ enabled?
+            disable_irq(); // Disable it
+            dmb();
+            int_state.cpu_irq_prev_interrupt_state.set(1);
+        } else {
+            // Make sure the to save the prev state as false
+            int_state.cpu_irq_prev_interrupt_state.set(0);
+        }
+    }
+
+    int_state
+        .cpu_irq_critical_section_counter
+        .set(int_state.cpu_irq_critical_section_counter.get() + 1);
+    atomic::compiler_fence(Ordering::Acquire);
+}
+
+/// Leave IRQ critical section
+fn cpu_irq_leave_critical(int_state: &InterruptState) {
+    atomic::compiler_fence(Ordering::Acquire);
+
+    // Check if the user is trying to leave a critical section
+    // when not in a critical section
+    if int_state.cpu_irq_critical_section_counter.get() > 0 {
+        int_state
+            .cpu_irq_critical_section_counter
+            .set(int_state.cpu_irq_critical_section_counter.get() - 1);
+
+        // Only enable global interrupts when the counter
+        // reaches 0 and the state of the global interrupt flag
+        // was enabled when entering critical state */
+        if int_state.cpu_irq_critical_section_counter.get() == 0
+            && int_state.cpu_irq_prev_interrupt_state.get() == 1
+        {
+            dmb();
+            enable_irq();
+        }
+    }
+
+    atomic::compiler_fence(Ordering::Acquire);
+}
+
+impl<'desc> DMA<'desc> {
+    /// Initialize a new DMA context
+    pub fn new(dmac: DMAC, descriptor_list: &'desc DescriptorList) -> Self {
+        Self {
+            dmac,
+            channel: None, // Channel not yet allocated
+            job_status: VolatileCell::new(Status::Ok),
+            has_descriptors: false, // No descriptors allocated yet
+            loop_flag: false,
+            peripheral_trigger: TRIGSRCW::DISABLE, // Software trigger only by default
+            trigger_action: TRIGACTW::TRANSACTION,
+            callbacks: Callbacks::default(),
+            descriptor_list,
+        }
+    }
+
+    // TODO(adafruit): Drop impl? Should stop job, delete descriptors, free channel.
+
+    /// IRQ handler
+    pub fn irq_handler(&mut self, flags: u8) {
```

Comment:
ZeroDMA has a global interrupt handler which looks up which channel received an interrupt. I think we'll need one too... this is a janky one that will only work for one channel (if that!).